src/MerkleGen.sol

```diff
@@ -272,4 +272,134 @@ library MerkleGen {
 
         return (proof_hashes, proof_flag_bits_bool, root);
     }
+
+    function gen(bytes32[] memory leaves, uint256 leafIndex) internal pure returns (bytes32[] memory, bytes32) {
+        require(leaves.length > 1, "MerkleGen: Leaves should be greater than 1.");
+
+        // Append with the same leaf if odd number of leaves
+        if (leaves.length % 2 == 1) {
+            leaves = leaves.append(leaves[leaves.length - 1]);
+        }
+
+        bytes32[] memory proof = get_proof(leaves, leafIndex);
+        bytes32 root = get_root(leaves);
+
+        return (proof, root);
+    }
+
+    function init_tree(bytes32[] memory leaves) internal pure returns (bytes32[] memory) {
+        require(leaves.length > 1, "MerkleGen: Leaves should be greater than 1.");
+
+        bytes32[] memory tree = new bytes32[](2 * leaves.length - 1);
+
+        uint256 index = tree.length - leaves.length;
+
+        for (uint256 i = 0; i < leaves.length; i++) {
+            tree[index + i] = leaves[i];
+        }
+
+        return tree;
+    }
+
+    function build_tree(bytes32[] memory leaves) internal pure returns (bytes32[] memory) {
+        bytes32[] memory tree = init_tree(leaves);
+
+        for (uint256 i = tree.length - 1; i > 1; i -= 2) {
+            bytes32 left = tree[i - 1];
+            bytes32 right = tree[i];
+            bytes32 parent = hash_internal_nodes(left, right);
+            uint256 parentIndex = (i - 1) / 2;
+            tree[parentIndex] = parent;
+        }
+
+        return tree;
+    }
+
+    function get_root(bytes32[] memory leaves) internal pure returns (bytes32) {
+        require(leaves.length > 1, "MerkleGen: Data should be greater than 1.");
+
+        bytes32[] memory tree = build_tree(leaves);
+
+        return tree[0];
+    }
+
+    function get_proof(bytes32[] memory leaves, uint256 index) internal pure returns (bytes32[] memory) {
```

Comment:
Function names in solidity should all be camelCase. It looks like we haven't followed this in the previous iterations either. It'd be good if you could update all of the functions to camelCase :pray:

Reply:
Changed all function names to camelCase.

Reply:
Thank you, they're looking good!